feat: start adding socket tests

mhdawson opened the pull request.

mhdawson commented:
@ospencer this is the start of adding tests for sock_accept. I've commented out the part that does the actual sock_accept call since that is not implemented yet but it did already find one issue on the pre-open side of things.

Looking at the method a bit more since our discussion I think that possibly the `on_new_connection` method can simply call `uv_cond_signal` on a condition associated with the socket.  Inside of sock_accept if the flags indicated that the call should be blocking then if an initial call to  uv_accept returns UV_EAGAIN then it would call `uv_cond_wait` on that same condition associated with the sock.

I think that implementation for `on_new_connection` would mean that if previous blocking call to sock_accept had been made before the connection came in it would be woken up when a new accept is made, otherwise if no connection has been made when the sock_accept is called it will either return with UV_EAGAIN if the flags indicated it should be non-blocking or block waiting until the connection later came in.  We'd also have to have some clean to be able to break out of the wait I think for shutdown.

mhdawson commented:
@ospencer I experimented a bit based on the discussion with @cjihrig  in https://github.com/nodejs/uvwasi/issues/211#issuecomment-1601615580 and added an initial implementation for uvwasi_sock_accept.  Based on the uvwasi documentation I'm not sure we can depend on how calls to uv_accept are being used but it seems to pass initial test test when the connection is made before the call. Next thing will be to test when the connection comes in after the uv_accept call. The test needs an extra thread to do and I ran out of time before being able to add that.

mhdawson commented:
:( CI seems to fail on mac os. I wonder if it might be because I've already closed the connection and I'm just getting `lucky` on linux. @ospencer any change you have a mac and can try/debug a bit there?

mhdawson commented:
I pushed all of the tests so the client runs off thread and so that there is a delay after the connectionis made before the connection close, earlier tests passed on macos as well :). New tests were added to check the cases where the connection comes in after a blocking call and that's looking good a well. No tests actually test that we can send/receive on the socket we get from uvwasi_sock_accept but so far looking good.

mhdawson commented:
@ospencer just so you are in the loop. Have also added initial implementations of uvwasi_sock_send and uvwasi_sock_receive.  Right now using global vars etc. so not what we'll want in the end but shows how using the async libuv calls looks like.

cjihrig commented:
@mhdawson is this ready for review?

mhdawson commented:
@cjihrig it still needs some work before it might be ready to land.  I've just found an issue that I'm current investigating while adding code in the test to close sockets, not sure if its a timing issue in the tests or a problem doing uv_accept. A few more things I think should happen before it is ready to land include:
* look through other calls that accept fds and see if any are required for minimal functionality (for example closing sockets). Implement socket support for minimal set, and for the others possibly add code to return a reasonable error value.
* cleanup pass

Having said that a top level scan by yourself to see it's heading in the right direction/using reasonable approaches would be useful to avoid too much more work if the work so far is going in the wrong direction.

mhdawson commented:
@cjihrig good news is that the accept issue was a problem in the uvwasi_sock_accept versus something in how uv_accept works :). Pushed a commit to fix.

One question, the libuv APIs don't seem to support closing the Read (versus write) part of the socket. Do you think we should throw a non-implemented error or simply add a flag to fd structure to mark that the socket has been closed for read and then return a non-writable error if reads are attempted on the socket later on.

cjihrig commented on src/fd_table.c:
```diff
@@ -307,7 +313,7 @@ uvwasi_errno_t uvwasi_fd_table_insert_preopen_socket(uvwasi_t* uvwasi,
                                 NULL,
                                 UVWASI_FILETYPE_SOCKET_STREAM,
-                                UVWASI__RIGHTS_SOCKET_BASE,
+                                UVWASI__RIGHTS_SOCKET_BASE | UVWASI_RIGHT_SOCK_ACCEPT,
```
Should `UVWASI_RIGHT_SOCK_ACCEPT` be part of `UVWASI__RIGHTS_SOCKET_BASE`? I'm not sure how much it matters because I believe the whole rights system has gone away in future versions of WASI.

mhdawson replied:
Added UVWASI_RIGHT_SOCK_ACCEPT be part of UVWASI__RIGHTS_SOCKET_BASE

cjihrig commented on test/test-enotsup-apis.c:
```diff
@@ -2,11 +2,6 @@
 #include "uvwasi.h"
 
 int main(void) {
-  /* TODO(cjihrig): This test is intended to be temporary. */
```
I guess we should either remove this file now or leave a new comment that this test is intended to verify the behavior of unimplemented APIs but currently there are none.

mhdawson replied:
Removed

cjihrig commented on test/test-sock-send-recv.c:
```diff
@@ -0,0 +1,162 @@
+#define INVALID_SOCK 42
+#define DEFAULT_BACKLOG 5
+
+int delayedThreadTime =  5000;
```
Prefer snake case over camelCase.

```suggestion
static int delayed_thread_time = 5000;
```

cjihrig commented on test/test-sock-send-recv.c:
```diff
@@ -0,0 +1,162 @@
+#define DEFAULT_BACKLOG 5
+
+int delayedThreadTime =  5000;
+int immedateThreadTime =  0;
```
```suggestion
static int immediate_thread_time = 0;
```

cjihrig commented on test/test-sock-send-recv.c:
```diff
@@ -0,0 +1,162 @@
+int immedateThreadTime =  0;
+
+#define CONNECT_ADDRESS "127.0.0.1"
+#define TEST_PORT_1 10500
```
Would it be possible to use port 0 to avoid conflicts?

cjihrig commented on test/test-sock-send-recv.c:
```diff
@@ -0,0 +1,162 @@
+}
+
+int main(void) {
+#if !defined(_WIN32) && !defined(__ANDROID__)
```
Not sure about Android, but shouldn't this work on Windows?

mhdawson replied:
Thanks for catching that, it was a copy paste error, I have no reason to believe they won't work on both. Wil remove the guard.

cjihrig commented on test/test-sock-send-recv.c:
```diff
@@ -0,0 +1,162 @@
+#define CONNECT_ADDRESS "127.0.0.1"
+#define TEST_PORT_1 10500
+
+void on_uv_close(uv_handle_t* handle) {
```
I think you can make these functions static.

cjihrig commented on test/test-sock-send-recv.c:
```diff
@@ -0,0 +1,162 @@
+
+void on_client_connect(uv_connect_t * req, int status) {
+  if (status < 0) {
+      fprintf(stderr, "New connection error %s\n", uv_strerror(status));
+      // error!
+      return;
```
Two space indentation please. Also, should this fail the test?

mhdawson replied:
Fixed the indentation. The test will fail if the data is not returned so I think fail wise its covered.  I was not sure if an assert in another thread will do the right thing. If so adding one would make sense.

cjihrig commented on test/test-sock-send-recv.c:
```diff
@@ -0,0 +1,162 @@
+
+void client_connection_echo_thread(void* time) {
+  int r = 0;
+  uv_loop_t *loop = malloc(sizeof(uv_loop_t));
```
I think you could reduce the number of `malloc()` and `free()` calls by creating static variables for the test.

cjihrig commented on test/test-sock-accept.c:
```diff
@@ -0,0 +1,149 @@
+#define INVALID_SOCK 42
+
+#define CONNECT_ADDRESS "127.0.0.1"
+#define TEST_PORT_1 10500
```
Some of the same questions/comments here as the other test:

- Use static data more.
- Use snake_case.
- Is it possible to use port 0 for the test.
- What about Windows?

cjihrig commented on test/test-sock-accept.c:
```diff
@@ -0,0 +1,149 @@
+
+void on_client_connect(uv_connect_t * req, int status) {
+  if (status < 0) {
+      fprintf(stderr, "New connection error %s\n", uv_strerror(status));
+      // error!
+      return;
```
Two space indentation and possibly fail the test if an error occurs.

cjihrig commented on src/uvwasi.c:
```diff
@@ -2593,30 +2699,187 @@ uvwasi_errno_t uvwasi_sock_send(uvwasi_t* uvwasi,
+  return UVWASI_ESUCCESS;
+}
+
+typedef struct shutdown_data_s {
```
Can we move some of these structs and helper functions to a separate file.

mhdawson replied:
Done

cjihrig commented on src/uvwasi.c:
```diff
@@ -2593,30 +2699,187 @@ uvwasi_errno_t uvwasi_sock_send(uvwasi_t* uvwasi,
+
+      int r = uv_accept((uv_stream_t*) wrap->sock, (uv_stream_t*) uv_connect_sock);
+      if (r == UV_EAGAIN) {
+	// still no connection or error so run the look again
```
```suggestion
	// still no connection or error so run the loop again
```

cjihrig commented on src/uvwasi.c:
```diff
@@ -2593,30 +2699,187 @@ uvwasi_errno_t uvwasi_sock_send(uvwasi_t* uvwasi,
+      }
+
+      if (r != 0) {
+	// and error occured accepting the connection breakout of the loop and
```
```suggestion
	// An error occurred accepting the connection. Break out of the loop and
```

cjihrig commented on src/uvwasi.c:
```diff
@@ -2593,30 +2699,187 @@ uvwasi_errno_t uvwasi_sock_send(uvwasi_t* uvwasi,
+
+      if (r != 0) {
+	// and error occured accepting the connection breakout of the loop and
+	// report and error
```
```suggestion
	// report an error.
```

mhdawson commented:
@cjihrig accepted changes/pushed a commit to fix the 'easy' review comments. Now looking at the rest.

mhdawson commented on src/fd_table.c:
```diff
@@ -74,35 +74,41 @@ uvwasi_errno_t uvwasi_fd_table_insert(uvwasi_t* uvwasi,
+  size_t mp_len = 0;
   char* mp_copy;
-  size_t rp_len;
+  size_t rp_len = 0;
```
Done

mhdawson commented:
The tests look like they are likely still timing out on Windows and MacOS.

I had created a Windows machine and they run fine on my local machine so not sure what is going on since the action does not seem to show any output until they actually time out.

mhdawson commented:
@cjihrig I think this is now ready for another review with the goal of getting it to lond.

The tests now run and pass on all of the CI platforms.

I believe I've addressed the comments except for using port 0. I'm not sure how we'd get the port that was used based on the available APIs.  

I've also added validation so that we indicate when a requested option is not supported as well as the standard debugging which was missing before.

cjihrig commented on test/test-sock-send-recv.c:
```diff
@@ -0,0 +1,168 @@
+
+  uv_sleep(*((int*) time));
+
+  uv_tcp_t* socket = (uv_tcp_t*)malloc(sizeof(uv_tcp_t));
```
I think you can avoid `malloc()` and `free()` for this and the `uv_connect_t` a few lines down. For example, see how [this libuv test does it](https://github.com/libuv/libuv/blob/d09441ca03e399fe641da88624c8ea0476967187/test/test-tcp-alloc-cb-fail.c#L29-L35).

cjihrig commented on test/test-sock-accept.c:
```diff
@@ -0,0 +1,156 @@
+#define CONNECT_ADDRESS "127.0.0.1"
+#define TEST_PORT_1 10500
+#define CONNECTION_WAIT_TIME 2000
+int delayedThreadTime = 4000;
+int immediateThreadTime = 0;
```
Please use snake case instead of camelCase.

mhdawson replied:
Pushed commit to fix that.

cjihrig commented on test/test-sock-accept.c:
```diff
@@ -0,0 +1,156 @@
+  free(handle);
+}
+
+void on_client_connect(uv_connect_t * req, int status) {
```
```suggestion
void on_client_connect(uv_connect_t* req, int status) {
```

cjihrig commented on test/test-sock-accept.c:
```diff
@@ -0,0 +1,156 @@
+      // error!
+      return;
+  }
+  uv_close((uv_handle_t *)req->handle, on_uv_close);
```
```suggestion
  uv_close((uv_handle_t*)req->handle, on_uv_close);
```

cjihrig reviewed (approved):
Great work. Left some comments, but mostly just style things that can be applied from the GitHub UI. After merging this, it's probably worth cutting a new release.

mhdawson commented on test/test-sock-accept.c:
```diff
@@ -0,0 +1,156 @@
+
+  uv_sleep(*((int*) time));
+
+  uv_tcp_t* socket = (uv_tcp_t*)malloc(sizeof(uv_tcp_t));
```
I think the difference is that this test uses multiple threads. There maybe be two threads using `  client_connection_thread` at the same time. 

For example in one part of the test we have

```C++
  // validate two accepts queue up properly
  start_client_connection_thread(&immediateThreadTime);
  start_client_connection_thread(&immediateThreadTime);
  start_client_connection_thread(&immediateThreadTime);
```

That applies to this case. For the other test we currenlty only have one instance but I think we should make it possible to have multiple launched to support future testing.
